packages/odyssey-react-mui/src/labs/FileUpload.tsx

```diff
@@ -0,0 +1,335 @@
+  acceptedFileTypes,
+  errorMessage,
+  id,
+  isButtonOnly,
+  isDisabled = false,
+  isOptional,
+  hint,
+  HintLinkComponent,
+  label,
+  onChange,
+  type,
+}: FileUploadProps) => {
+  const odysseyDesignTokens = useOdysseyDesignTokens();
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const [filesToUpload, setFilesToUpload] = useState<File[]>([]);
+
+  useEffect(() => {
+    onChange(filesToUpload);
+  }, [filesToUpload, onChange]);
+
+  const handleFileInputChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      const { files } = event.target;
+
+      if (files && files.length > 0) {
+        const mergedFiles = type ? [...filesToUpload, ...files] : [...files];
+
+        setFilesToUpload(mergedFiles);
+      }
+
+      // reset input value to allow re-upload of a file with the same name
+      event.target.value = "";
+    },
+    [type, filesToUpload]
+  );
+
+  const handleUploadButtonClick = useCallback(() => {
+    inputRef?.current?.click();
+  }, [inputRef]);
+
+  const handleFileRemoval = useCallback(
+    (name: string) => {
+      const deletedFileFilteredOut = filesToUpload.filter(
+        (file) => file.name != name
+      );
+      setFilesToUpload(deletedFileFilteredOut);
+    },
+    [filesToUpload]
+  );
+
+  const renderFileInput = useCallback(
+    ({ ariaDescribedBy, errorMessageElementId, id, labelElementId }) => {
+      const fileNames = filesToUpload.map((file) => file.name);
+
+      const Input = () => (
+        <input
+          accept={acceptedFileTypes?.toString()}
+          aria-describedby={ariaDescribedBy}
+          aria-errormessage={errorMessageElementId}
+          aria-labelledby={labelElementId}
+          disabled={isDisabled}
+          id={id}
+          multiple={type === "multiple"}
+          onChange={handleFileInputChange}
+          ref={inputRef}
+          title=""
+          type="file"
+        />
+      );
+
+      if (isButtonOnly) {
+        return (
+          <>
+            <BaseInputWrapper>
+              <Input />
+              <Button
+                isDisabled={isDisabled}
+                label="Upload Files"
+                onClick={handleUploadButtonClick}
+                startIcon={<UploadIcon />}
+                variant="secondary"
+              />
+            </BaseInputWrapper>
+            <FileUploadPreview
+              fileNames={fileNames}
+              handleFileRemoval={handleFileRemoval}
+              isDisabled={isDisabled}
+            />
+          </>
+        );
+      }
+
+      return (
+        <>
+          <InputContainer odysseyDesignTokens={odysseyDesignTokens}>
+            <Input />
+            <ButtonAndInfoContainer>
+              <UploadIllustration />
+              <Support color="textSecondary">
+                Drag and drop files here or click to add files.
+              </Support>
+              <Button
+                isDisabled={isDisabled}
+                label="Add Files"
```

Comment:
Does this need translations?

Reply:
Yeah, we do. Hadn't added that in yet. Good call